aws.go

```diff
@@ -0,0 +1,439 @@
+func (a *AWS) getNodes(network string) (*nodes, error) {
+	ns, err := a.common.getNodes(network)
+	if err != nil {
+		return nil, err
+	}
+	if ns == nil {
+		err = a.refresh()
+		if err != nil {
+			return nil, err
+		}
+		ns, err = a.common.getNodes(network)
+	}
+	return ns, err
+}
+
+// SetNode inserts or updates the node.
+func (a *AWS) setNode(node *node) error {
+	err := a.setNodeSecrets(node)
+	if err != nil {
+		return err
+	}
+	item := NodeItem{
+		node.network,
+		node.domain,
+		node.created,
+		node.expires,
+		node.role,
+		node.scrambler.key}
+
+	av, err := dynamodbattribute.MarshalMap(item)
+	if err != nil {
+		fmt.Println("Got error marshalling new creator item:")
+		return err
+	}
+
+	input := &dynamodb.PutItemInput{
+		Item:      av,
+		TableName: aws.String(nodesTableName),
+	}
+
+	_, err = a.svc.PutItem(input)
+	if err != nil {
+		fmt.Println("Got error calling PutItem:")
+		return err
+	}
+
+	return nil
+}
+
+func (a *AWS) refresh() error {
+	nets := make(map[string]*nodes)
+
+	// Fetch the nodes and then add the secrets.
+	ns, err := a.fetchNodes()
+	if err != nil {
+		return err
+	}
+	err = a.addSecrets(ns)
+	if err != nil {
+		return err
+	}
+
+	// Create a map of networks from the nodes found.
+	for _, v := range ns {
+		net := nets[v.network]
+		if net == nil {
+			net = &nodes{}
+			net.dict = make(map[string]*node)
+			nets[v.network] = net
+		}
+		net.all = append(net.all, v)
+		net.dict[v.domain] = v
+	}
+
+	// Finally sort the nodes by hash values and whether they are active.
+	for _, net := range nets {
+		net.order()
+	}
+
+	// In a single atomic operation update the reference to the networks and
+	// nodes.
+	a.mutex.Lock()
+	a.nodes = ns
+	a.networks = nets
+	a.mutex.Unlock()
+
+	return nil
+}
+
+func (a *AWS) fetchNodes() (map[string]*node, error) {
+	var err error
+	ns := make(map[string]*node)
+
+	params := &dynamodb.ScanInput{
+		TableName: aws.String(nodesTableName),
+	}
+
+	// Fetch all the records from the nodes table in Dynamo.
+	result, err := a.svc.Scan(params)
+	if err != nil {
+		fmt.Println("Query API call failed:")
+		fmt.Println((err.Error()))
+		return nil, err
+	}
+
+	// Iterate over the records creating nodes and adding them to the networks
+	// map.
+	for _, i := range result.Items {
+		nodeItem := NodeItem{}
+
+		err = dynamodbattribute.UnmarshalMap(i, &nodeItem)
+		if err != nil {
+			fmt.Println("Got error unmarshalling:")
+			fmt.Println(err.Error())
+			return nil, err
+		}
+
+		ns[nodeItem.Domain], err = newNode(
+			nodeItem.Network,
+			nodeItem.Domain,
+			nodeItem.Created,
+			nodeItem.Expires,
+			nodeItem.Role,
+			nodeItem.ScramblerKey)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return ns, err
+}
+
+func (a *AWS) addSecrets(ns map[string]*node) error {
+
+	// Fetch all the records from the secrets table in Azure.
+	params := &dynamodb.ScanInput{
+		TableName: aws.String(nodesTableName),
+	}
+
+	result, err := a.svc.Scan(params)
```

Comment:
Would it make sense to reuse the `fetchNodes` here?

Reply:
That should be scanning the secrets table.